waiter/src/waiter/scheduler/cook.clj

```diff
@@ -493,18 +497,25 @@
         search-interval (t/days search-interval-days)]
     (->CookScheduler service-id->password-fn service-id->service-description-fn
                      cook-api allowed-priorities allowed-users backend-port home-path-prefix
-                     search-interval service-id->failed-instances-transient-store)))
+                     search-interval service-id->failed-instances-transient-store retrieve-syncer-state-fn)))
 
 (defn cook-scheduler
   "Creates and starts cook scheduler with associated daemons."
-  [{:keys [failed-tracker-interval-ms http-options impersonate mesos-slave-port url] :as config}]
+  [{:keys [failed-tracker-interval-ms http-options impersonate mesos-slave-port url
+           scheduler-state-chan scheduler-syncer-interval-secs start-scheduler-syncer-fn] :as config}]
+  {:pre [(not (nil? scheduler-state-chan))
+         (utils/pos-int? scheduler-syncer-interval-secs)
+         (not (nil? start-scheduler-syncer-fn))]}
   (let [http-client (http-utils/http-client-factory http-options)
         cook-api {:http-client http-client
                   :impersonate impersonate
                   :slave-port mesos-slave-port
                   :spnego-auth (:spnego-auth http-options false)
                   :url url}
         service-id->failed-instances-transient-store (atom {})
-        scheduler (create-cook-scheduler config cook-api service-id->failed-instances-transient-store)]
+        syncer-state-atom (atom {})
+        retrieve-syncer-state-fn (partial scheduler/retrieve-syncer-state)
```

Comment:
Wraps the access to the atom and hence limits the writers of the atom. We cannot initialize the atom inside the syncer as there is a circular dependency between the syncer and the scheduler. Using the atom allows us to break the circular dependency.